### ISSUE TITLE
watir-webdriver + firefox is very slow to loop table rows

### ISSUE THREAD

**Issue by alexhanh** [watir/watir-webdriver#328]
I'm on OS X 10.10.2. I've tried all the possible combinations of these versions:

watir-webdriver: 0.6.11, 0.7.0
Firefox: 33.03, 34.0.5, 36.0.4

but regardless of the versions, the following loop is painfully slow:

`browser.table(id: 'table').tbody.trs.each { |tr| puts tr.id }`
and
`browser.table(id: 'table').tbody.trs.to_a.each { |tr| puts tr.id }`

20 rows takes about 20 seconds to loop like this.

Is it possible to make this loop faster? What is causing it to be so slow? Thanks!

**Comment by titusfortner** [comment]
Yes, unfortunately this is a Selenium bug in version 2.45- https://code.google.com/p/selenium/issues/detail?id=8551

Firefox before version 36 and using selenium-webdriver 2.45.0.dev3 should work

**Comment by alexhanh** [comment]
@titusfortner Ah, thanks. Are you saying watir-webdriver is currently unusable or how does the selenium-webdriver fix this (I'm unfamiliar with selenium-webdriver)? Are we supposed to simply include the selenium-webdriver in the Gemfile? Thanks again!

**Comment by alexhanh** [comment]
To answer my own question: add the following to your `Gemfile`

`gem 'selenium-webdriver', '2.45.0.dev3'`

**Comment by alexhanh** [comment]
After downgrading to FF34 and selenium-webdriver 2.45.0.dev3, the loop still takes 10 seconds.

Is there a way to force a "prefetch" so that the information would be cached and not piped into the browser everytime?

Right now I'm thinking doing this:

`doc = Nokogiri::HTML(browser.table(id: 'table').html)`

and using Nokogiri to scrape the information in much faster manner. Am I think along the right lines here? Thanks!

**Comment by titusfortner** [comment]
How fast does it take you in Chrome?
Do you notice a difference between 0.6.11 and 0.7? (we changed some things that could affect speed)

**Comment by titusfortner** [comment]
Also, @jkotests has this project already, based on @zeljkofilipin's idea: https://jkotests.wordpress.com/2013/08/21/watir-nokogiri-gem-published/ 

I kind of think it would be interesting to bake this into Watir itself in a way that doesn't require a duplicate project. I see Watir more as a testing framework than a WebDriver DSL. As such, Watir shouldn't *need to use WebDriver for things that would be much faster/better with an HTML library.

I'm more focused right now on improving Watir's reliability as a test framework than digging into improving performance, but it is an interesting idea.

**Comment by p0deje** [comment]
There is a big difference between using nokogiri and watir: with nokogiri you can retrieve any data you need (+ there are tons of other stuff for that like mechanize), with watir - you get only what end user gets. For instance, if element is invisible, `#text` will return empty string which is the same as user sees, while nokogiri would still returns non-empty string.

Having a separate `watir-nokogiri` or `watir-mechanize` seems like a great idea: it's a sort of new [celeriy](https://github.com/jarib/celerity).

**Comment by alexhanh** [comment]
Thank you guys for this enlightening discussion.

@titusfortner Unfortunately 0.7.0 requires selenium-webdriver (>= 2.45), so I couldn't test 0.7.0 with 2.45.0.dev3. 0.6.11 with 2.45.0.dev3 definitely feels much faster than 0.7.0 with 2.45.0 (but that could very well be to the mentioned Selenium bug).

I can totally understand the benefits of watir being always in sync with the real browser state (especially important in end-to-end testing among other things). I would argue, however, that from what I've personally experienced and seen other people go through, providing a watir-nokogiri like experience would make sense right in this gem. The performance issues seem to be relevant a lot of times. The tradeoffs would be needed to be documented very clearly, sure.

So +1 for bringing watir-nokogiri implementation to watir-webdriver. Thanks again!

**Comment by alexhanh** [comment]
I just realized that `tbody` and `trs` in a chain like `browser.table(id: 'table').tbody.trs` probably traverses the whole DOM sub-tree instead of just immediate children. If the sub-trees are deep and large, then it's not that surprising that these kind of queries are slow.

I'm curious to experiment what effect would limiting the search to immediate children have. Is there a way to restrict the search to just immediate children (would you need to revert to using css/xpath selectors) and do you think this could have meaningful effect?

**Comment by p0deje** [comment]
What do you mean by traversing the whole sub-tree vs immediate children?

The chain is the same as locating all elements with CSS `table#table tbody tr`.

**Comment by alexhanh** [comment]
I think this author does a lot better job explaining what I'm trying to say http://www.natontesting.com/2009/08/19/how-to-get-only-direct-child-objects-in-watir/

My example was pretty poor (well, unless the table uses nested tables).

**Comment by p0deje** [comment]
I see what you mean. You can still use XPath/CSS to retrieve first childs: 

```ruby
browser.trs(xpath: "//table[@id='table']/tbody/tr")
```

I would say that in most cases you should just limit the locating elements (e.g. by adding class selector), I'd argue that retrieving the first child is more expected behavior - current implementation is much more flexible (since allows to use both approaches).

**Comment by alexhanh** [comment]
Cool, that works if we begin from the document root and/or have an absolute reference point. But often times we are searching relative to a Watir Element (because we need the parent element as well as the children). So, how would you conveniently limit the search scope to the immediate children in this case?

```
table = browser.table(id: 'table')
table.trs.each do |tr| # problem, not immediate
  tr.tds.each do |td| # problem, not immediate
    # ...
  end 
end
```

**Comment by p0deje** [comment]
I would just limit the lookup by adding classes to trs/tds. 

On the other hand, I see the problem you are talking about even though I have personally never had any problems with it. I see always locating the immediate child has a lot of problems (for example, you have to follow the whole chain) but it would be great other people thoughts on this.

**Comment by Blackbaud-TitusFortner** [comment]
I've been working a lot on deeply nested elements recently. 0.6.11 actually does not support certain forms of nesting at all. This was fixed for 0.7.0, but at the expense of some performance. My current pull request fixes that performance significantly for iterating over deeply nested elements, but it shouldn't affect performance issues you might be having with 0.6.11.

A quick example of how Watir will process this in the next version:

    trs = browser.table.trs.to_a
    trs.each do |tr| 
      tds = tr.tds.to_a
      tds.each do |td|
        puts td.text
      end
    end

browser.table.trs.to_a locates all of the trs on the page under the table with wire calls like this:
1) Verify window isn't closed --> getWindowHandles & getCurrentWindowHandle
2) Ensure driver is in top level browsing context --> switchToFrame
3) Look up table --> findElement
4) Verify Selenium Element is a table --> getElementTagName
5) Look up all trs --> findChildElements

For each tr, tr.td.to_a does:
1) Verify parent isn't stale --> isElementEnabled
2) Look up child elements --> findChildElements

For each td.text:
1) Verify td exists --> isElementEnabled
2) Get text --> getElementText

Total wire calls:
6+2*trs.size+2*(total_number_tds)

So, there aren't any superfluous calls to the driver here, as each call gives important information to the overall test.

I have not worked with xpath selectors much. I think you should still be able to nest them the same way:
browser.table(id: 'table').trs(xpath: "/tbody/tr").each { |tr| tr.tds(xpath: "/td"). each { |td| puts td.text} }

It's implemented with: @wd.find_elements(:xpath, XPATH_STRING)[idx]
and @wd would be the Selenium::WebDriver::Element of the parent, so it should only look at specific children of that element each call.

Are you noticing the same performance issue in Chrome? It could be a limitation for the Firefox driver itself, which will hopefully be reduced when Mozilla finally releases Marionette.

**Comment by alexhanh** [comment]
Switching to Chrome didn't seem to have any significant impact.

Using `trs(xpath: './*') and `tds(xpath: './*')` was slower than just `trs` and `tds`.

`tbody.trs(xpath: '/tr')` returns no matches (the selector must be wrong).

I got it working by using `tbody.trs(xpath: './tr')` and `tr.tds(xpath: './td')', but this also didn't have any significant impact it seems.

Hmm, would be pretty neat if we could have a standard test website to benchmark against.
